File: app/Models/ExternalActivityRequest.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class ExternalActivityRequest extends Model
{
    /**
     * Max hours a student may request across external activity requests
     * in a single academic year (pending + approved, rejected excluded).
     */
    public const ANNUAL_HOUR_CAP = 10;

    protected $fillable = [
        'user_id',
        'title',
        'organization',
        'activity_date',
        'activity_category',
        'hours_requested',
        'hours_approved',
        'proof_image_path',
        'status',
        'reject_reason',
        'admin_comment',
        'reviewed_by',
        'reviewed_at',
    ];

    protected function casts(): array
    {
        return [
            'activity_date' => 'date',
            'reviewed_at' => 'datetime',
        ];
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function reviewer(): BelongsTo
    {
        return $this->belongsTo(User::class, 'reviewed_by');
    }

    public function scopePending($query)
    {
        return $query->where('status', 'pending');
    }

    /**
     * The hours actually credited: what an admin overrode to, or the
     * student's original request if never adjusted.
     */
    public function getHoursCreditedAttribute(): int
    {
        return $this->hours_approved ?? $this->hours_requested;
    }

    /**
     * Hours already claimed by a student against a given academic year's
     * ANNUAL_HOUR_CAP, counting pending + approved requests (rejected
     * requests free up the quota). $dateColumn selects whether the
     * academic year is judged by when the activity happened
     * (activity_date) or when it was submitted (created_at).
     */
    public static function hoursUsedInAcademicYear(int $userId, int $academicYear, string $dateColumn): int
    {
        [$start, $end] = \App\Services\AcademicYearCalculator::rangeFor($academicYear);

        return (int) static::where('user_id', $userId)
            ->whereIn('status', ['pending', 'approved'])
            ->whereBetween($dateColumn, [$start, $end])
            ->selectRaw('COALESCE(SUM(COALESCE(hours_approved, hours_requested)), 0) as total')
            ->value('total');
    }
}

File: mobile/lib/core/models/external_activity_request.dart
class ExternalActivityRequest {
  ExternalActivityRequest({
    required this.id,
    required this.title,
    required this.organization,
    required this.activityDate,
    required this.activityCategory,
    required this.hoursRequested,
    this.hoursApproved,
    required this.hoursCredited,
    required this.status,
    this.rejectReason,
  });

  final int id;
  final String title;
  final String organization;
  final String activityDate;
  final String activityCategory;
  final int hoursRequested;
  final int? hoursApproved;
  final int hoursCredited;
  final String status;
  final String? rejectReason;

  factory ExternalActivityRequest.fromJson(Map<String, dynamic> json) {
    return ExternalActivityRequest(
      id: json['id'] as int,
      title: json['title'] as String,
      organization: json['organization'] as String,
      activityDate: json['activity_date'] as String,
      activityCategory: json['activity_category'] as String,
      hoursRequested: json['hours_requested'] as int,
      hoursApproved: json['hours_approved'] as int?,
      hoursCredited: json['hours_credited'] as int,
      status: json['status'] as String,
      rejectReason: json['reject_reason'] as String?,
    );
  }
}
